### PR TITLE
gh-112069: Make sets thread-safe with the GIL disabled

### PR DESCRIPTION
**Pull request by tomasr8**
Still have some failing tests in the GIL version so marking as a draft for now until I manage to fix them.
This PR is mostly based on Sam's version: https://github.com/colesbury/nogil-3.12/commit/4ca2924f0d

Since sets have a lot of methods that are exposed, I made this list to keep track and help with the review:

## set methods
| Method | Thread-safe (comments) |
|---|---|
| set_add                                   | ✅
| set_clear                                 | ✅
| set.\_\_contains__                       | ✅
| set_copy                                  | ✅
| set_discard                               | ✅
| set_difference_multi                      | ✅
| set_difference_update                     | ✅
| set_intersection_multi                    | ✅
| set_intersection_update_multi             | ✅
| set_isdisjoint                            | ✅
| set_issubset                              | ✅
| set_issuperset                            | ✅
| set_pop                                   | ✅
| set.\_\_reduce__                                | ✅
| set_remove                                | ✅
| set.\_\_sizeof__                                | ✅
| set_symmetric_difference                  | ✅
| set_symmetric_difference_update | ✅
| set_union                                 | ✅
| set_update                                | ✅

## `set_as_number`
| Method | Thread-safe (comments) |
|---|---|
| set_sub  | ✅
| set_and  | ✅
| set_xor  | ✅ (via `set_symmetric_difference`)
| set_or   | ✅
| set_isub | ✅
| set_iand | ✅
| set_ixor | ✅
| set_ior  | ✅

## set `tp_`..
| Method | Thread-safe (comments) |
|---|---|
| set_dealloc        | Not needed
| set_repr           | ✅
| set_as_sequence(set_len, set_contains)    | ✅
| set_traverse       | Not needed
| set_clear_internal | Not needed
| set_richcompare    | ❌ (will be done separately)
| set_iter           | ✅
| set_init           | ✅
| set_new            | ✅
| set_vectorcall     | ✅

There are also some extra frozenset methods which I don't think need locking: `frozenset_copy`, `frozenset_hash`, `frozenset_new`, `frozenset_vectorcall`. The rest of frozenset methods are shared with sets.

## C API methods:
| Method | Thread-safe (comments) |
|---|---|
| PySet_New        | ✅
| PyFrozenSet_New  | ✅
| PySet_Size       | ✅ (via `set_len`)
| PySet_Clear      | ✅ (via `set_clear`)
| PySet_Contains   | ✅
| PySet_Discard    | ✅
| PySet_Add        | ✅
| _PySet_NextEntry |  ✅
| PySet_Pop        | ✅ (via `set_pop`)
| _PySet_Update    | ✅

* Issue: gh-112069

### CODE REVIEW

File: Objects/setobject.c [hunk 1]
```diff
@@ -488,8 +516,9 @@ static void
 set_dealloc(PySetObject *so)
 {
     setentry *entry;
-    Py_ssize_t used = so->used;
+    Py_ssize_t used;
 
+    used = so->used;
```
Comment (colesbury) [inline Objects/setobject.c]: Revert unnecessary change:

```suggestion
    Py_ssize_t used = so->used;
```

File: Objects/setobject.c [hunk 2]
```diff
@@ -615,32 +657,75 @@ set_merge(PySetObject *so, PyObject *otherset)
...
         }
     }
-    return 0;
+
+exit:
+    Py_END_CRITICAL_SECTION2();
+    return rv;
+}
+
+static int set_merge_dict(PySetObject *so, PyObject *other) {
```
Comment (colesbury) [inline Objects/setobject.c]: CPython coding style:

```suggestion
static int
set_merge_dict(PySetObject *so, PyObject *other)
{
```

File: Objects/setobject.c [hunk 3]
```diff
@@ -1181,106 +1255,132 @@ set_ior(PySetObject *so, PyObject *other)
 }
 
 static PyObject *
-set_intersection(PySetObject *so, PyObject *other)
-{
-    PySetObject *result;
-    PyObject *key, *it, *tmp;
+set_intersection_set(PySetObject *so, PySetObject *other) {
```
Comment (colesbury) [inline Objects/setobject.c]:
```suggestion
set_intersection_set(PySetObject *so, PySetObject *other)
{
```

File: Objects/setobject.c [hunk 4]
```diff
@@ -1344,124 +1448,163 @@ set_iand(PySetObject *so, PyObject *other)
     return Py_NewRef(so);
 }
 
+static PyObject *
+set_isdisjoint_set(PySetObject *so, PySetObject *other) {
```
Comment (colesbury) [inline Objects/setobject.c]:
```suggestion
set_isdisjoint_set(PySetObject *so, PySetObject *other)
{
```

File: Objects/setobject.c [hunk 5]
```diff
@@ -1181,106 +1255,132 @@ set_ior(PySetObject *so, PyObject *other)
...
 set_intersection_multi(PySetObject *so, PyObject *args)
 {
     Py_ssize_t i;
+    PyObject *result;
 
     if (PyTuple_GET_SIZE(args) == 0)
         return set_copy(so, NULL);
 
-    PyObject *result = Py_NewRef(so);
+    Py_BEGIN_CRITICAL_SECTION(so);
```
Comment (colesbury) [inline Objects/setobject.c]: We should not have a critical section here as the locking is done in the call to `set_intersection()`.

**Review by colesbury — COMMENTED** [review]
Hi @tomasr8, thanks so much for working on this!

First, please let me know if you would like help debugging the test failures.

One of the challenges with this is that it's hard to know at a glance if a function does locking internally or assumes that the caller is supposed to do the locking. I think this will be easier to maintain if we generally push the locking to the outermost functions. Argument clinic makes this a bit easier by generating some of the critical section calls. I think we should:

1) Convert the functions to use argument clinic with the `@critical_section` directive.
2) For C APIs (like `PySet_Add()`) put the critical section code in the outermost function.

Sorry, this suggestion wasn't mentioned in the issue -- support for `@critical_section` in Argument Clinic was added a few days after I wrote up the set issue.

At the top of the file:

```
/*[clinic input]
class set "PySetObject *" "&PySet_Type"
[clinic start generated code]*/
/*[clinic end generated code: output=da39a3ee5e6b4b0d input=abe13a1b24961902]*/

#include "clinic/setobject.c.h"
```

And then, `set_add` would look like:

```c

/*[clinic input]
@critical_section
set.add

    key: object
    /

Add an element to a set.

This has no effect if the element is already present.
[clinic start generated code]*/

static PyObject *
set_add_impl(PySetObject *self, PyObject *key)
/*[clinic end generated code: output=8d849b1bd2bd8b3a input=33fb8030b1ad21f5]*/
{
    if (set_add_key(self, key))
        return NULL;
    Py_RETURN_NONE;
}
```

**Comment by colesbury** [comment]
On the thread-safety table:

Do not need to be thread-safe: `set_dealloc `, `set_traverse`, `set_clear_internal` (i.e., `tp_clear` hook).

`set_richcompare` should be thread-safe, but we may want to defer and just make a note of it
`_PySet_NextEntry` needs to be addressed separately due to borrowed reference counts
`set_iter` should be thread-safe. I don't think it needs locking, just load the size once via an atomic op.

**Comment by tomasr8** [comment]
Thanks a lot for the review! I wasn't aware of the new AC directive - I'll update the PR. As for the test failures, I haven't had the time yet to really dig into it, hopefully I'll have the time tomorrow. If I can't figure it out myself, I'll let you know :)

**Comment by erlend-aasland** [comment]
This is growing to be a large diff. I think we should consider splitting this up in two PRs:

1. adapt the needed methods to Argument Clinic
2. make it thread safe

**Comment by tomasr8** [comment]
> This is growing to be a large diff. I think we should consider splitting this up in two PRs:
> 
>     1. adapt the needed methods to Argument Clinic
> 
>     2. make it thread safe

Makes sense. I'll open a separate PR for the AC changes once I figure out how to convert all the methods :)

**Comment by erlend-aasland** [comment]
> I'll open a separate PR for the AC changes once I figure out how to convert all the methods :)

Great :) You can open a draft PR and ping me if you need help!

**Comment by tomasr8** [comment]
I've adapted the PR to use `@critical_section` wherever I could. Everywhere else I tried to put the critical sections in the outermost functions.

The only problem I still have is that the nogil build fails when I use the critical section macros in the C API functions:
```bash
./configure --with-pydebug --disable-gil --config-cache
make -s -j2
Segmentation fault (core dumped)
make: *** [Makefile:1607: Python/frozen_modules/getpath.h] Error 139
make: *** Waiting for unfinished jobs....
Segmentation fault (core dumped)
make: *** [Makefile:1612: Python/frozen_modules/importlib._bootstrap.h] Error 139
```
I'm guessing I might've forgotten some import or to regenerate some file? `getpath.h` is rather cryptic as well. What's weird is that I can build it just fine with the GIL enabled..

**Comment by colesbury** [comment]
Hi @tomasr8, the crash is because `PySet_New(iterable)` accepts a `NULL` iterable, but `Py_BEGIN_CRITICAL_SECTION(iterable)` does not work with a `NULL` argument and will crash. Here's how I debugged it:

1) The `make` output indicated that the `./Programs/_freeze_module` invocations were crashing. `_freeze_module` is a C program that embeds Python. 
2) I ran `gdb --args ./Programs/_freeze_module getpath ./Modules/getpath.py Python/frozen_modules/getpath.h` to re-run the command under GDB
3) I typed `run` to run the program and then, after the Segmentation Fault, I ran `bt` (for backtrace) 
4) The first three lines of the backtrace look like:

```
#1  PyMutex_LockFast (lock_bits=0xa <error: Cannot access memory at address 0xa>) at ./Include/internal/pycore_lock.h:61
#2  _PyCriticalSection_Begin (m=0xa, c=0x7fffffffd980) at ./Include/internal/pycore_critical_section.h:179
#3  PySet_New (iterable=iterable@entry=0x0) at Objects/setobject.c:2543
```
5) The first thing that stood out to me was `m=0xa`, which doesn't look like a valid pointer. These small but not quite zero pointers are common when you dereference a field from a NULL pointer. The second thing I noticed was `iterable=iterable@entry=0x0`, which is the same as `NULL`.

I'll add some suggestions on the PR.

File: Objects/setobject.c [hunk 6]
```diff
@@ -2408,13 +2538,23 @@ PyTypeObject PyFrozenSet_Type = {
 PyObject *
 PySet_New(PyObject *iterable)
 {
-    return make_new_set(&PySet_Type, iterable);
+    PyObject *rv;
+
+    Py_BEGIN_CRITICAL_SECTION(iterable);
```
Comment (colesbury) [inline Objects/setobject.c]: We don't want a critical section here. In the cases where we need to lock `iterable`, we should make that decision deeper down when we know more about the type of `iterable`.

File: Objects/setobject.c [hunk 7]
```diff
@@ -2408,13 +2538,23 @@ PyTypeObject PyFrozenSet_Type = {
...
+    return rv;
 }
 
 PyObject *
 PyFrozenSet_New(PyObject *iterable)
 {
-    return make_new_set(&PyFrozenSet_Type, iterable);
+    PyObject *rv;
+
+    Py_BEGIN_CRITICAL_SECTION(iterable);
```
Comment (colesbury) [inline Objects/setobject.c]: Same here -- no critical section.

File: Objects/setobject.c [hunk 8]
```diff
@@ -2184,21 +2310,25 @@ set_vectorcall(PyObject *type, PyObject * const*args,
...
-    set_len,                            /* sq_length */
+    (lenfunc)set_len,                            /* sq_length */
     0,                                  /* sq_concat */
     0,                                  /* sq_repeat */
     0,                                  /* sq_item */
     0,                                  /* sq_slice */
     0,                                  /* sq_ass_item */
     0,                                  /* sq_ass_slice */
-    (objobjproc)set_contains,           /* sq_contains */
+    (objobjproc)set_sq_contains,           /* sq_contains */
```
Comment (colesbury) [inline Objects/setobject.c]: The comments should be aligned with each other. Or, alternatively, this could be written to use C99 designated initializers:

```
static PySequenceMethods set_as_sequence = {
    .sq_length = (lenfunc)set_len,
    .sq_contains = (objobjproc)set_sq_contains.
};
```

File: Objects/setobject.c [hunk 9]
```diff
@@ -2184,21 +2310,25 @@ set_vectorcall(PyObject *type, PyObject * const*args,
     }
 
     if (nargs) {
-        return make_new_set(_PyType_CAST(type), args[0]);
+        PyObject *rv;
+        Py_BEGIN_CRITICAL_SECTION(args[0]);
+        rv = make_new_set(_PyType_CAST(type), args[0]);
+        Py_END_CRITICAL_SECTION();
+        return rv;
```
Comment (colesbury) [inline Objects/setobject.c]: I don't think we want to lock here.

File: Objects/setobject.c [hunk 10]
```diff
@@ -2424,50 +2564,75 @@ PySet_Size(PyObject *anyset)
...
     if (!PySet_Check(set)) {
         PyErr_BadInternalCall();
         return -1;
     }
-    return set_clear_internal((PySetObject *)set);
+
+    Py_BEGIN_CRITICAL_SECTION(set);
+    rv = set_clear_internal((PySetObject *)set);
+    Py_END_CRITICAL_SECTION();
+    return rv;
```
Comment (colesbury) [inline Objects/setobject.c]: This is fine, but you can also call `set_clear((PySetObject *)set, NULL);`, which does the locking internally and so doesn't need extra `Py_BEGIN_CRITICAL_SECTION`.

File: Objects/setobject.c [hunk 11]
```diff
@@ -1833,14 +1926,19 @@ set_ixor(PySetObject *so, PyObject *other)
 
     if (!PyAnySet_Check(other))
         Py_RETURN_NOTIMPLEMENTED;
+
+    Py_BEGIN_CRITICAL_SECTION2(so, other);
     result = set_symmetric_difference_update(so, other);
+    Py_END_CRITICAL_SECTION2();
```
Comment (colesbury) [inline Objects/setobject.c]: `set_symmetric_difference_update` (from setobject.c.h) already has a critical section.

File: Objects/setobject.c [hunk 12]
```diff
@@ -1821,9 +1907,16 @@ set_symmetric_difference(PySetObject *so, PyObject *other)
...
 {
+    PyObject *rv;
+
     if (!PyAnySet_Check(so) || !PyAnySet_Check(other))
         Py_RETURN_NOTIMPLEMENTED;
-    return set_symmetric_difference(so, other);
+
+    Py_BEGIN_CRITICAL_SECTION2(so, other);
+    rv = set_symmetric_difference(so, other);
+    Py_END_CRITICAL_SECTION2();
```
Comment (colesbury) [inline Objects/setobject.c]: `set_symmetric_difference` (from setobject.c.h) already has a critical section.

File: Objects/setobject.c [hunk 13]
```diff
@@ -1663,17 +1722,27 @@ set_difference_multi_impl(PySetObject *so, PyObject *args)
     Py_ssize_t i;
     PyObject *result, *other;
 
-    if (PyTuple_GET_SIZE(args) == 0)
-        return set_copy(so, NULL);
+    if (PyTuple_GET_SIZE(args) == 0) {
+        Py_BEGIN_CRITICAL_SECTION(so);
+        result = set_copy(so, NULL);
+        Py_END_CRITICAL_SECTION();
```
Comment (colesbury) [inline Objects/setobject.c]: `set_copy(so, NULL);` already has a critical section (`set_copy_impl` does not)

File: Objects/setobject.c [hunk 14]
```diff
@@ -1188,15 +1202,21 @@ set_union_impl(PySetObject *so, PyObject *args)
     PyObject *other;
     Py_ssize_t i;
 
+    Py_BEGIN_CRITICAL_SECTION(so);
     result = (PySetObject *)set_copy(so, NULL);
+    Py_END_CRITICAL_SECTION();
```
Comment (colesbury) [inline Objects/setobject.c]: `set_copy()` already has a critical section

File: Objects/setobject.c [hunk 15]
```diff
@@ -1207,31 +1227,43 @@ set_union_impl(PySetObject *so, PyObject *args)
 static PyObject *
 set_or(PySetObject *so, PyObject *other)
 {
-    PySetObject *result;
+    PySetObject *result = NULL;
 
     if (!PyAnySet_Check(so) || !PyAnySet_Check(other))
         Py_RETURN_NOTIMPLEMENTED;
 
+    Py_BEGIN_CRITICAL_SECTION2(so, other);
```
Comment (colesbury) [inline Objects/setobject.c]: `set_copy()` already has a critical section

File: Objects/setobject.c [hunk 16]
```diff
@@ -1333,13 +1365,21 @@ set_intersection_multi_impl(PySetObject *so, PyObject *args)
 {
     Py_ssize_t i;
 
-    if (PyTuple_GET_SIZE(args) == 0)
-        return set_copy(so, NULL);
+    if (PyTuple_GET_SIZE(args) == 0) {
+        PyObject *result;
+        Py_BEGIN_CRITICAL_SECTION(so);
+        result = set_copy(so, NULL);
```
Comment (colesbury) [inline Objects/setobject.c]: `set_copy` already has a critical section

**Review by colesbury — COMMENTED** [review]
This is looking good overall. 

- Consider adding `_Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(obj);` to functions where you assume that the caller locks the object. For example, `set_update_internal` would be a good candidate
- Dino added a few critical sections here that should probably be removed now that you are doing locking in the callers to those functions: in `set_update_internal` (for the dict) and `set_symmetric_difference_update`.
- Since `iterable` may be NULL, do the critical section inside `make_new_set` instead of the callers to `make_new_set`

File: Objects/setobject.c (hunk 8 above)
Comment (rhettinger) [reply Objects/setobject.c]: Just align the comments and keep the PR focused on its core task and minimize the diff as much as possible.  Changing variable names, rewriting logic, or making stylistic changes to long stable code is just a distractor and will make the edit harder to review.

**Comment by tomasr8** [comment]
> Hi @tomasr8, the crash is because `PySet_New(iterable)` accepts a `NULL` iterable, but `Py_BEGIN_CRITICAL_SECTION(iterable)` does not work with a `NULL` argument and will crash. Here's how I debugged it:
> 
>     1. The `make` output indicated that the `./Programs/_freeze_module` invocations were crashing. `_freeze_module` is a C program that embeds Python.
> 
>     2. I ran `gdb --args ./Programs/_freeze_module getpath ./Modules/getpath.py Python/frozen_modules/getpath.h` to re-run the command under GDB
> 
>     3. I typed `run` to run the program and then, after the Segmentation Fault, I ran `bt` (for backtrace)
> 
>     4. The first three lines of the backtrace look like:
> 
> 
> ```
> #1  PyMutex_LockFast (lock_bits=0xa <error: Cannot access memory at address 0xa>) at ./Include/internal/pycore_lock.h:61
> #2  _PyCriticalSection_Begin (m=0xa, c=0x7fffffffd980) at ./Include/internal/pycore_critical_section.h:179
> #3  PySet_New (iterable=iterable@entry=0x0) at Objects/setobject.c:2543
> ```
> 
>     5. The first thing that stood out to me was `m=0xa`, which doesn't look like a valid pointer. These small but not quite zero pointers are common when you dereference a field from a NULL pointer. The second thing I noticed was `iterable=iterable@entry=0x0`, which is the same as `NULL`.
> 
> 
> I'll add some suggestions on the PR.

Amazing analysis! This helps me a lot, thanks :)

File: Objects/setobject.c (hunk 6 above)
Comment (tomasr8) [reply Objects/setobject.c]: I moved the locking inside `make_new_set`, lmk if it's ok like that :)

Comment (colesbury) [reply Objects/setobject.c]: Perfect

File: Objects/setobject.c [hunk 17]
```diff
@@ -889,13 +895,14 @@ PyTypeObject PySetIter_Type = {
...
         return NULL;
     si->si_set = (PySetObject*)Py_NewRef(so);
-    si->si_used = so->used;
+    si->si_used = size;
     si->si_pos = 0;
-    si->len = so->used;
+    si->len = size;
     _PyObject_GC_TRACK(si);
     return (PyObject *)si;
 }
```
Comment (colesbury) [inline Objects/setobject.c:918]: In the `set_update_internal` function below, we should:

1) Add assertions that `so` and `other` are locked (the assertions also allow refcount == 1).
 ```
    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(so);
    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(other);
```

2) Remove the `Py_BEGIN_CRITICAL_SECTION`/`Py_END_CRITICAL_SECTION` in the dict case. Now that `set` performs the locking we don't want this nested critical section. It's both slow (because re-acquiring an already held critical section is slow) and may be incorrect (because only the inner-most critical sections is guaranteed to be held, it will unlock the set `so`).

File: Objects/setobject.c (hunk 6 above)
Comment (colesbury) [reply Objects/setobject.c]: Sorry, I'm realizing I gave bad advice. This approach leads to a lot of redundant locking of the same object can be quite slow. For example, `set_copy()` locks `so` but that's locked again in `make_new_set` (as `iterable`).

I think the previous approach in which the locking is in the outer functions will work better, but it will have to handle `NULL` values in some cases. We should probably add a variant of `Py_BEGIN_CRITICAL_SECTION` that handles (and ignores) `NULL`

File: Objects/setobject.c (hunk 17 above)
Comment (tomasr8) [reply Objects/setobject.c:918]: I added the assertions to `set_update_internal` and removed the locking for dictionaries. Other candidates for assertions could also be `set{add,contains,discard}key`

**Comment by tomasr8** [comment]
To give an update, now that `Py_XBEGIN_CRITICAL_SECTION` can be used, I just need to move the locking from `make_new_set` to `PySet_New` and `PyFrozenSet_New`.

However there is a small issue which is that `set_update_internal` (called by `make_new_set`) requires both its arguments to be locked. I could simply keep the locking in `make_new_set` but as @colesbury mentioned that would be quite inefficient. I think the assertions in `set_update_internal` are useful so I'm trying to come up with a solution where we can keep them but also avoid locking in `make_new_set`.

**Comment by colesbury** [comment]
Hi @tomasr8, thanks so much for the work you've done in making set thread-safe. The remaining pieces are a bit tricky, and I don't think I can give helpful advice without actually making the edits, so I'm going to make some changes directly to the PR.

File: Objects/setobject.c [hunk 18]
```diff
@@ -889,58 +904,59 @@ PyTypeObject PySetIter_Type = {
...
-        Py_hash_t hash;
-        Py_ssize_t dictsize = PyDict_GET_SIZE(other);
+    /* Do one big resize at the start, rather than
+    * incrementally resizing as we insert new keys.  Expect
+    * that there will be no (or few) overlapping keys.
+    */
+    Py_ssize_t dictsize = PyDict_GET_SIZE(other);
+    if ((so->fill + dictsize)*5 >= so->mask*3) {
+        if (set_table_resize(so, (so->used + dictsize)*2) != 0)
+            return -1;
```
Comment (erlend-aasland) [inline Objects/setobject.c]:
```suggestion
        if (set_table_resize(so, (so->used + dictsize)*2) != 0) {
            return -1;
        }
```

File: Objects/setobject.c [hunk 19]
```diff
@@ -955,6 +971,73 @@ set_update_internal(PySetObject *so, PyObject *other)
...
+{
+    if (PyAnySet_Check(other)) {
+        return set_merge_lock_held(so, other);
+    }
+    else if (PyDict_CheckExact(other)) {
+        return set_update_dict_lock_held(so, other);
+    }
+    else {
+        return set_update_iterable_lock_held(so, other);
+    }
```
Comment (erlend-aasland) [inline Objects/setobject.c]:
```suggestion
    return set_update_iterable_lock_held(so, other);
```

File: Objects/setobject.c [hunk 20]
```diff
@@ -955,6 +971,73 @@ set_update_internal(PySetObject *so, PyObject *other)
...
+    else if (PyDict_CheckExact(other)) {
+        int rv;
+        Py_BEGIN_CRITICAL_SECTION(other);
+        rv = set_update_dict_lock_held(so, other);
+        Py_END_CRITICAL_SECTION();
+        return rv;
+    }
+    else {
+        return set_update_iterable_lock_held(so, other);
+    }
```
Comment (erlend-aasland) [inline Objects/setobject.c]:
```suggestion
    return set_update_iterable_lock_held(so, other);
```

File: Objects/setobject.c [hunk 21]
```diff
@@ -955,6 +971,73 @@ set_update_internal(PySetObject *so, PyObject *other)
...
+}
+
+static int
+set_update_internal(PySetObject *so, PyObject *other)
+{
+    if (PyAnySet_Check(other)) {
+        if ((PyObject *)so == other) {
+            return 0;
+        }
```
Comment (erlend-aasland) [inline Objects/setobject.c]:
```suggestion
        if (Py_Is((PyObject *)so, other)) {
            return 0;
        }
```

File: Objects/setobject.c [hunk 22]
```diff
@@ -1213,11 +1308,13 @@ set_or(PySetObject *so, PyObject *other)
         Py_RETURN_NOTIMPLEMENTED;
 
     result = (PySetObject *)set_copy(so, NULL);
-    if (result == NULL)
+    if (result == NULL) {
         return NULL;
-    if ((PyObject *)so == other)
+    }
+    if ((PyObject *)so == other) {
```
Comment (erlend-aasland) [inline Objects/setobject.c]:
```suggestion
    if (Py_Is((PyObject *)so, other)) {
```

File: Objects/setobject.c [hunk 23]
```diff
@@ -1380,17 +1482,26 @@ set_intersection_update_multi_impl(PySetObject *so, PyObject *args)
...
     set_swap_bodies(so, (PySetObject *)tmp);
+    Py_END_CRITICAL_SECTION();
     Py_DECREF(tmp);
     Py_RETURN_NONE;
 }
 
 static PyObject *
 set_and(PySetObject *so, PyObject *other)
 {
+    PyObject *rv;
```
Comment (erlend-aasland) [inline Objects/setobject.c]: I'd consider moving this declaration down to just above L1500, for a smaller diff.

File: Objects/setobject.c [hunk 24]
```diff
@@ -2434,7 +2614,8 @@ PySet_Clear(PyObject *set)
         PyErr_BadInternalCall();
         return -1;
     }
-    return set_clear_internal((PySetObject *)set);
+    set_clear((PySetObject *)set, NULL);
```
Comment (erlend-aasland) [inline Objects/setobject.c]:
```suggestion
    (void)set_clear((PySetObject *)set, NULL);
```

File: Objects/setobject.c (hunk 23 above)
Comment (erlend-aasland) [reply Objects/setobject.c]: Ditto for `set_sub` and `set_isub`.

File: Objects/setobject.c [hunk 25]
```diff
@@ -1684,46 +1815,89 @@ set_difference_multi_impl(PySetObject *so, PyObject *args)
...
+{
+    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(so);
+    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(other);
+
+    Py_ssize_t pos = 0;
+    setentry *entry;
+    while (set_next(other, &pos, &entry)) {
+        PyObject *key = entry->key;
+        Py_hash_t hash = entry->hash;
+        Py_INCREF(key);
```
Comment (erlend-aasland) [inline Objects/setobject.c]:
```suggestion
        PyObject *key = Py_NewRef(entry->key);
        Py_hash_t hash = entry->hash;
```

File: Objects/setobject.c [hunk 26]
```diff
@@ -1739,58 +1913,41 @@ static PyObject *
...
 {
-    PySetObject *otherset;
-    PyObject *key;
-    Py_ssize_t pos = 0;
-    Py_hash_t hash;
-    setentry *entry;
-    int rv;
-
-    if ((PyObject *)so == other)
+    if ((PyObject *)so == other) {
```
Comment (erlend-aasland) [inline Objects/setobject.c]:
```suggestion
    if (Py_Is((PyObject *)so, other)) {
```

**Review by erlend-aasland — COMMENTED** [review]
Some style nitpicks.

File: Objects/setobject.c [hunk 27]
```diff
@@ -567,14 +569,24 @@ set_repr(PySetObject *so)
...
+    Py_END_CRITICAL_SECTION();
+    return result;
+}
+
 static Py_ssize_t
-set_len(PyObject *so)
+set_len(PySetObject *so)
 {
-    return ((PySetObject *)so)->used;
+    return _Py_atomic_load_ssize_relaxed(&so->used);
```
Comment (DinoV) [inline Objects/setobject.c]: This could be `FT_ATOMIC_LOAD_SSIZE_RELAXED(so->used)` now

File: Objects/setobject.c [hunk 28]
```diff
@@ -2156,13 +2323,17 @@ set_init(PySetObject *self, PyObject *args, PyObject *kwds)
 {
     PyObject *iterable = NULL;
 
-     if (!_PyArg_NoKeywords("set", kwds))
+    if (!_PyArg_NoKeywords("set", kwds))
         return -1;
     if (!PyArg_UnpackTuple(args, Py_TYPE(self)->tp_name, 0, 1, &iterable))
         return -1;
+
+    Py_BEGIN_CRITICAL_SECTION(self);
```
Comment (DinoV) [inline Objects/setobject.c:2332]: Maybe it's worth inlining `set_update_internal` into here?  It'd mean we'd only take the critical section once, and also so that the clear and update are done atomically.  As of now someone could call `__init__` on multiple threads and end up with elements from both callers.

**Review by DinoV — APPROVED** [review]

File: Objects/setobject.c (hunk 28 above)
Comment (colesbury) [reply Objects/setobject.c:2332]: It's a bit more complicate than just inlining `set_update_internal` because sometimes we want to lock both `self` and `iterable`, sometimes `iterable` does not need to be locked, and sometimes `iterable` is NULL. Additionally, we would like to avoid locking `self` if it's a newly created and not yet visible, which is the most common case.

I'd like to address this, but in a subsequent PR.

**Comment by tomasr8** [comment]
> Hi @tomasr8, thanks so much for the work you've done in making set thread-safe. The remaining pieces are a bit tricky, and I don't think I can give helpful advice without actually making the edits, so I'm going to make some changes directly to the PR.

Go for it :) Sorry for the lack of activity, I've been sick for the last couple of weeks :/

**Comment by colesbury** [comment]
@tomasr8, I hope you feel better soon and no need to apologize. I really appreciate the work you've done on set and previously with making hashlib thread-safe.

**Comment by colesbury** [comment]
Hmmm... the failures are really confusing. It's unclear to me if they are related to this PR. So far, I don't think I've seen failures on other buildbots.
